File: LeetCode/Hard/1255-maximum-score-words-formed-by-letters/1255-maximum-score-words-formed-by-letters.py
class Solution:
    def maxScoreWords(self, words: List[str], letters: List[str], score: List[int]) -> int:
        n = len(words)
        ret = [0]
        count = Counter(letters)
        
        def dfs(idx, curr, counter):
            if idx == n:
                ret[0] = max(ret[0], curr)
                return
            
            count = Counter(words[idx])
            possible = True
            
            for k, v in count.items():
                if counter[k] < v:
                    possible = False
                
            if possible:
                temp = 0
                for k, v in count.items():
                    temp += score[ord(k) - 97] * v
                    counter[k] -= v

                dfs(idx + 1, curr + temp, counter)
                
                for k, v in count.items():
                    counter[k] += v
                
                dfs(idx + 1, curr, counter)
            else:
                dfs(idx + 1, curr, counter)
        
        dfs(0, 0, count)
        
        return ret[0]
                

File: LeetCode/Hard/1255-maximum-score-words-formed-by-letters/1255-maximum-score-words-formed-by-letters.go
func maxScoreWords(words []string, letters []byte, score []int) int {
    n := len(words)
    ret := 0
    wordsCount := make([][]int, n)
    count := make([]int ,26)
    
    for i := 0; i < n; i++ { wordsCount[i] = make([]int, 26) }
    
    for i := 0; i < n; i++ {
        for _, v := range words[i] {
            wordsCount[i][v-'a']++
        }
    }
    
    for i := 0; i < len(letters); i++ { count[letters[i] - 'a']++ }

    dfs := func(idx, curr int){}
    dfs = func(idx, curr int){
        if idx == n {
            ret = max(ret, curr)
            return
        }
        
        currWord := wordsCount[idx]
        possible := true

        for i := 0; i < 26; i++ {
            if currWord[i] > count[i]{
                possible = false
                break
            }
        }

        if possible {
            temp := 0
            for i := 0; i < 26; i++ {
                temp += score[i] * currWord[i]
                count[i] -= currWord[i]
            }

            dfs(idx + 1, curr + temp)

            for i := 0; i < 26; i++ {
                count[i] += currWord[i]
            }

            dfs(idx + 1, curr)
        } else { dfs(idx + 1, curr) }
    }

    dfs(0, 0)

    return ret
}

func max (a, b int) int {
    if a > b { return a }
    return b
}